### PR TITLE
Fix: Display correct author dates when birth/death years contain non-numeric characters

### LINKED ISSUE #10352: Librarian-only: Empty brackets after author name
### Problem

OL displays the years of birth and death in author links if you are signed in as a librarian. When both fields are empty, this is omitted. But when the fields contain something that no year can be parsed from, an empty pair of brackets with a hyphen is displayed.

[Example page](https://openlibrary.org/books/OL22062526M/China) where the author’s birth year cannot be parsed because I set it to “19xx” to indicate that she was born in the 20th century.
Image

### Context

- Browser (Chrome, Safari, Firefox, etc): Safari
- OS (Windows, Mac, etc):
- Logged in (Y/N): Y
- Environment (prod, dev, local): prod

#### Requirements Checklist
* [ ]

#### Related files

* https://github.com/internetarchive/openlibrary/blob/d9fb9438d420953c6a1ef1442fbd8127abf40b4e/openlibrary/macros/BookByline.html#L7-L12

#### Stakeholders

*
<hr>

#### Instructions for Contributors

- Please [run these commands](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#working-on-your-branch) to ensure your repository is up to date **before** [creating a new branch](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#making-changes-and-creating-a-pull-request) to work on this issue and **each time after** pushing code to Github, because the pre-commit bot may add commits to your PRs upstream.

### PR DESCRIPTION
**Pull request by SharkyBytes**
Closes #10352 

Fix: Display correct author dates when birth/death years contain non-numeric characters

### Technical

-> Updated the `extract_year` method in `helpers.py` to handle non-numeric characters in birth/death dates.
-> Modified the `BookByline.html` template to ensure proper display of dates, even when non-numeric characters (e.g., "19xx") are present.
-> Added an optional `int_only` parameter to the `extract_year` method to avoid breaking changes in other parts of the codebase.

### Testing

->  Verify that the author's dates are displayed correctly without empty brackets.
->  Run the test suite to ensure no regressions:

### Screenshot

What I changed -:
image

### Stakeholders

@cdrini

### CODE REVIEW

File: openlibrary/core/helpers.py [hunk 1]
```diff
@@ -335,13 +335,19 @@ def private_collection_in(collections: list[str]) -> bool:
     return any(x in private_collections() for x in collections)
 
 
-def extract_year(input: str) -> str:
-    """Extracts the year from an author's birth or death date."""
-    if result := re.search(r'\d{4}', input):
-        return result.group()
-    else:
+def extract_year(input: str, int_only: bool = True) -> str:
+
```
Comment (cdrini) [inline openlibrary/core/helpers.py:339]: Let's bring the comment back in here, no reason to delete it

File: openlibrary/core/helpers.py [hunk 2]
```diff
@@ -335,13 +335,19 @@ def private_collection_in(collections: list[str]) -> bool:
...
+    if not input:
         return ''
 
+    if match := re.search(r'\b(\d{4})\b', input):
+        return match.group(1)
+
+    if not int_only:
```
Comment (cdrini) [inline openlibrary/core/helpers.py:346]: Let's still keep this case strict, and have a new regex that _also_ allows non-int character. eg `[0-9xX?]` are some of the characters we often use for non-int years (eg `19xx` or `19??`).

File: openlibrary/core/helpers.py [hunk 3]
```diff
@@ -335,13 +335,19 @@ def private_collection_in(collections: list[str]) -> bool:
     return any(x in private_collections() for x in collections)
 
 
-def extract_year(input: str) -> str:
-    """Extracts the year from an author's birth or death date."""
-    if result := re.search(r'\d{4}', input):
-        return result.group()
-    else:
+def extract_year(input: str, int_only: bool = True) -> str:
```
Comment (cdrini) [inline openlibrary/core/helpers.py:338]: This method has become a bit more complicated, so lets add some unit tests to openlibrary/tests/core/test_helpers.py

File: openlibrary/macros/BookByline.html [hunk 4]
```diff
@@ -5,15 +5,19 @@
   $ url = author.get('url')
   $ rendered_name = cond(name, truncate(name, 40), _("Unknown author"))
   $if show_years and author.get('birth_date'):
```
Comment (cdrini) [inline openlibrary/macros/BookByline.html:7]: Let's store these out here; otherwise it's a little confused when these variables are defined. Also rename `birth_date_formatted` to `birth_year` to make it a little shorter.

```suggestion
  $ birth_year = ''
  $ death_year = ''
  $if show_years and author.get('birth_date'):
```

File: openlibrary/macros/BookByline.html [hunk 5]
```diff
@@ -5,15 +5,19 @@
...
-    $ birth_date_formatted = extract_year(author.get('birth_date'))
+    $ birth_date_formatted = extract_year(author.get('birth_date'), int_only=False)
     $if author.get('death_date'):
-      $ death_date_formatted = extract_year(author.get('death_date'))
+      $ death_date_formatted = extract_year(author.get('death_date'), int_only=False)
     $else:
       $ death_date_formatted = ''
-  $if show_years and url and author.get('birth_date'):
+  $if show_years and url and author.get('birth_date') and author.get('death_date'):
```
Comment (cdrini) [inline openlibrary/macros/BookByline.html:13]: Then here we can reference the `birth_year` and `death_year` variables directly instead of relying on `author.get('birth_date')` implying that `$birth_date_formatted` exists, which is a bit confusing.

```suggestion
  $if show_years and url and birth_year and death_year:
```

**Review by cdrini — CHANGES_REQUESTED** [review]
Awesome work @SharkyBytes ! A few pieces of feedback to make the logic a little clearer :+1:

File: openlibrary/core/helpers.py (hunk 2 above)
Comment (SharkyBytes) [reply openlibrary/core/helpers.py:346]: Ok @cdrini, got it! One doubt though—are we assuming the input will always be in this format? I’m allowing the last two characters to be non-int, and if they aren’t, it will reject the value. Is that fine? Like, It would allow values like '19XX', '20??', '18xX'.

File: openlibrary/macros/BookByline.html (hunk 4 above)
Comment (SharkyBytes) [reply openlibrary/macros/BookByline.html:7]: Okay, I also corrected the naming for this. Previously, it was birth_date_formatted, but now it has been changed to birth_year. Let me know @cdrini if the picture below looks good or not.

image

Comment (cdrini) [reply openlibrary/macros/BookByline.html:7]: Yep that looks good! Feel free to push it up, easier to comment on in code review 👍

Although double check that death_year clause, and also don't forget to add another one for when `url` is falsy!

Comment (cdrini) [reply openlibrary/macros/BookByline.html:7]: And actually this is statement is beginning to get a bit difficult to manage with an these cases, let's restructure like so:

Eg

```
$if url:
    <a href="..." $:attrs>
$else:
    <span $:attrs>

$rendered_name

$if show_years:
    $if birth_year and death_year:
        ($birth_year - $death_year)
    $elif birth_year:
        ($birth_year -)
    $elif death_year:
        (- $death_year)

$if url:
    </a>
$else:
    </span>
```

File: openlibrary/core/helpers.py (hunk 2 above)
Comment (cdrini) [reply openlibrary/core/helpers.py:346]: Technically the date spec does allow the X and ? Anywhere, so for simplicity I think using a regex like `[0-9xX?]{4}` might be best

File: openlibrary/macros/BookByline.html (hunk 4 above)
Comment (SharkyBytes) [reply openlibrary/macros/BookByline.html:7]:
```
   $elif death_year:
        (- $death_year)
```

Your overall structure looks quite modular, but @cdrini, do we really need this? Don’t you think it's a bit odd to have a death year? If we encounter a situation like this, only the name will be rendered.

Comment (SharkyBytes) [reply openlibrary/macros/BookByline.html:7]:
```
$def render_author_link(author):
  $ name = author.get('name')
  $ url = author.get('url')
  $ rendered_name = cond(name, truncate(name, 40), _("Unknown author"))

  $ birth_year = ''
  $ death_year = ''
  $if show_years and author.get('birth_date'):
    $ birth_year = extract_year(author.get('birth_date'), int_only=False)
  $if show_years and author.get('death_date'):
    $ death_year = extract_year(author.get('death_date'), int_only=False)
  
  $if url:
    <a href="$url" $:attrs>
  $else:
    <span $:attrs>

  $rendered_name

  $if show_years:
    $if birth_year and death_year:
      ($birth_year – $death_year)
    $elif birth_year:
      ($birth_year - )

  $if url:
    </a>
  $else:
    </span>
  ```
  
@cdrini, I've made this change for now. Let me know if you think any improvements are needed. Also, if we decide to include the death year, let me know, and I'll add it back into the code

Comment (cdrini) [reply openlibrary/macros/BookByline.html:7]: That seems reasonable to me!

Comment (cdrini) [reply openlibrary/macros/BookByline.html:7]: Actually, we can clean this up a bit more still! Now we have that tidy `$if show_years`, block, we can put the birth_year definitions/extractions inside that block, and kept that related code close together!

Comment (SharkyBytes) [reply openlibrary/macros/BookByline.html:7]: Great @cdrini, now I moved the birth year and death year definitions and extractions inside the `$if show_years` block. Have a look at the code now—I believe it's fully modular, finalized, and has reduced redundant, repetitive code. Still, let me know if you think there's any further scope for improvement.

```
$def render_author_link(author):
  $ name = author.get('name')
  $ url = author.get('url')
  $ rendered_name = cond(name, truncate(name, 40), _("Unknown author"))

  $if url:
    <a href="$url" $:attrs>
  $else:
    <span $:attrs>

  $rendered_name

  $if show_years:
    $ birth_year = ''
    $ death_year = ''
    $if author.get('birth_date'):
      $ birth_year = extract_year(author.get('birth_date'), int_only=False)
    $if author.get('death_date'):
      $ death_year = extract_year(author.get('death_date'), int_only=False)

    $if birth_year and death_year:
      ($birth_year – $death_year)
    $elif birth_year:
      ($birth_year - )
    $elif death_year:
      ( - $death_year)

  $if url:
    </a>
  $else:
    </span>
   ```

Comment (cdrini) [reply openlibrary/macros/BookByline.html:7]: Looks great! Push it up! 😊

Comment (cdrini) [reply openlibrary/macros/BookByline.html:7]: Actually, I think now we might be able to make a nice change to the ---

Nah just kidding 😁 I have no more feedback, it looks perfect!

Comment (SharkyBytes) [reply openlibrary/macros/BookByline.html:7]: Ah, Finally! Glad the changes look good😅. Now, I’d love to get your feedback on the other files as well—it would be great to refine their structure also and ensure everything is consistent.

File: openlibrary/core/helpers.py (hunk 2 above)
Comment (SharkyBytes) [reply openlibrary/core/helpers.py:346]:
```
def extract_year(input: str, int_only: bool = True) -> str:
    """Extracts the year from an author's birth or death date."""
    if not input:
        return '' 

    if int_only:
        if result := re.search(r'\b(\d{4})\b', input):
            return result.group(1)
    else:
        if result := re.search(r'\b([0-9xX?]{4})\b', input):
            return result.group(1)

    return ''

```

@cdrini :) I know there's still room for more modularity, but I feel the if and else statements are now pretty clear and make it easy to understand what the function is trying to achieve. Right?

That said, if you think we need to make it more precise or modular, let me know. Would love to get a review on this as well!

Comment (cdrini) [reply openlibrary/core/helpers.py:346]: Looks good!

File: openlibrary/macros/BookByline.html (hunk 4 above)
Comment (cdrini) [reply openlibrary/macros/BookByline.html:7]: All looks good so far! Once you do your next `git push` I'll give things another review, but I think you've addressed all my comments except for the unit tests one 👍

File: openlibrary/core/helpers.py (hunk 3 above)
Comment (SharkyBytes) [reply openlibrary/core/helpers.py:338]: Hey @cdrini,
The testing part is new for me, and I want to make sure I’m doing it the right way. I tried running all the tests, and the validations passed, but your validation matters a lot to me. 😊
Let me know if I added the tests incorrectly or if there's anything I should improve.
```
def test_extract_year():
    # Test numeric years
    assert h.extract_year("Born in 1980") == "1980"
    assert h.extract_year("Died in 2005") == "2005"
    assert h.extract_year("Year 1999") == "1999"
    
    # Test non-numeric years
    assert h.extract_year("Born in 19xx", int_only=False) == "19xx"
    assert h.extract_year("Died in 19??", int_only=False) == "19??"
    assert h.extract_year("Year 20XX", int_only=False) == "20XX"
    
    # Test no match
    assert h.extract_year("No year here") == ""
    assert h.extract_year("Year 123", int_only=False) == ""
  
 ``` 
  
image

File: openlibrary/core/helpers.py (hunk 2 above)
Comment (SharkyBytes) [reply openlibrary/core/helpers.py:346]: @cdrini,
I changed this file again. When I added the tests, the following assertion failed:

` assert h.extract_year("Died in 19??", int_only=False) == "19??"` 

Previously, I was using a boundary condition, but I’ve now removed it. I also checked the current code in the repo, and we’re not using boundary conditions in the regex pattern there either. So, I’ve kept the same behavior in this implementation as well.

Let me know if any changes are needed or if this is ready to pass your feedback.😊

```
def extract_year(input: str, int_only: bool = True) -> str:
    """Extracts the year from an author's birth or death date."""
    if not input:
        return '' 

    if int_only:
        pattern = r'\d{4}' 
    else:
        pattern = r'[0-9xX?]{4}'

    if result := re.search(pattern, input):
        return result.group()

    return ''

File: openlibrary/core/helpers.py (hunk 3 above)
Comment (cdrini) [reply openlibrary/core/helpers.py:338]: Yep that looks great, nice work! I've rarely seen text in the birth date field (like "born in 1923") it's usually just the date, so try some examples like "1923", "1923-03-14", "7 Dec 2018", "19xx". Those are more real-world 👍

File: openlibrary/core/helpers.py (hunk 2 above)
Comment (cdrini) [reply openlibrary/core/helpers.py:346]: Great catch! I was wondering about that change actually, so unit tests did the right thing! 😁

File: openlibrary/core/helpers.py (hunk 3 above)
Comment (SharkyBytes) [reply openlibrary/core/helpers.py:338]: Yeah, true @cdrini! I’ve updated it to this now!
```
def test_extract_year():
    assert h.extract_year("1980") == "1980"
    assert h.extract_year("2005-03-14") == "2005"
    assert h.extract_year("7 Dec 1999") == "1999"   

    assert h.extract_year("19xx", int_only=False) == "19xx"
    assert h.extract_year("6 Jan 19??", int_only=False) == "19??"
    assert h.extract_year("Year 20XX", int_only=False) == "20XX"
    
    assert h.extract_year("No year here") == ""
    assert h.extract_year("Year 123", int_only=False) == ""

File: openlibrary/core/helpers.py (hunk 2 above)
Comment (SharkyBytes) [reply openlibrary/core/helpers.py:346]: @cdrini Awesome! Glad the unit tests had my back on this one. Now I can finally give my keyboard a break😆Let me know if anything else pops up.

**Comment by SharkyBytes** [comment]
Hi @cdrini,
Just checking in— is there anything else I need to work on, or are you satisfied with the resolved changes? If any further adjustments are needed, please let me know, and I'll be happy to address them.

**Review by cdrini — APPROVED** [review]
Looks great! Made one small change to take advantage of the default return of `extract_year` being `''`, and tested on testing.openlibrary.org :

✅ Only has birth + inexact; has no birth or death year ([eg](https://testing.openlibrary.org/books/OL22062526M/China))
image

✅ Has both birth/death ([eg](https://testing.openlibrary.org/works/OL471576W/Murder_on_the_Orient_Express))

**Comment by SharkyBytes** [comment]
Thank you for merging @cdrini! Now, finally I can get back to my other issues and hopefully get ticked off on some 2025 checklist plans🙂
